### PR TITLE
Migrate Catch2 from v2 to v3

### LINKED ISSUE #246: Catch2 v3
Hello. I was trying to build `autodiff` locally where I have Catch2 v3 installed and realized that `autodiff` is still using Catch v2. Is there any interest or plan to migrate to Catch v3, now that it's been [officially released](https://github.com/catchorg/Catch2/releases/tag/v3.0.1)?

I did the migration on [my fork](https://github.com/RMeli/autodiff/tree/catch2-v3), and everything seems to work well. I can open a PR if this is of interest at all.

### PR DESCRIPTION
**Pull request by RMeli**
This PR migrates Catch2 from v2 to v3, following the [migration instructions](https://github.com/catchorg/Catch2/blob/devel/docs/migrate-v2-to-v3.md#top).

@allanleal Catch is also used in `test/BUILD`. I see in the [installation instructions](https://autodiff.github.io/installation/) that this is for the [Bazel](https://bazel.build/) build system. I never used it before and I can't see it in the CI builds; is there any documentation related to `autodiff` about this?

Fix #246

### CODE REVIEW

**Comment by allanleal** [comment]
Hi @RMeli , I work with CMake only. Bazel support was provided in the past by a contributor. There is no Bazel installation instructions at the moment (https://autodiff.github.io/installation/).

File: tests/reverse/var/var.test.cpp [hunk 1]
```diff
@@ -28,12 +28,13 @@
...
 // Catch includes
-#include <catch2/catch.hpp>
+#include <catch2/catch_approx.hpp>
+#include <catch2/catch_test_macros.hpp>
 
 // autodiff includes
 #include <autodiff/reverse/var.hpp>
 
-using autodiff::derivatives;
+    using autodiff::derivatives;
```
Comment (allanleal) [inline tests/reverse/var/var.test.cpp:37]:
```suggestion
using autodiff::derivatives;
```

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -1,16 +1,13 @@
...
+find_package(Catch2 3 REQUIRED)
 
 file(GLOB_RECURSE TEST_FILES "*.test.cpp")
 add_executable(autodiff-cpptests main.cpp ${TEST_FILES})
-target_link_libraries(autodiff-cpptests autodiff::autodiff Eigen3::Eigen Catch2::Catch2)
+target_link_libraries(autodiff-cpptests autodiff::autodiff Eigen3::Eigen Catch2::Catch2WithMain)
 set_target_properties(autodiff-cpptests PROPERTIES CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON CXX_EXTENSIONS OFF)
 
-# Add target tests that performs all C++ and Python tests
-add_custom_target(tests
```
Comment (allanleal) [inline tests/CMakeLists.txt:10]: We need the target `tests`, which is used in the CI (e.g., `ninja tests`). This code here removes it.

**Review by allanleal — CHANGES_REQUESTED** [review]

**Comment by RMeli** [comment]
> I work with CMake only. Bazel support was provided in the past by a contributor.

I see. I will try to wrap my head around it.

**Comment by RMeli** [comment]
@allanleal, is `bazel` integration updated at all since you work with CMake and it does not seem to be used in CI? I just tried to run `bazel test //...` (as suggested in https://github.com/autodiff/autodiff/pull/126#issuecomment-699948239) on the `main` branch, and I obtained the following error:
```
INFO: Analyzed 27 targets (2 packages loaded, 33 targets configured).
INFO: Found 4 targets and 23 test targets...
ERROR: /Users/rmeli/Documents/git/software/autodiff/examples/forward/BUILD:85:8: Compiling examples/forward/example-forward-gradient-derivatives-using-eigen-with-scalar.cpp failed: missing input file '//examples/forward:example-forward-gradient-derivatives-using-eigen-with-scalar.cpp'
ERROR: /Users/rmeli/Documents/git/software/autodiff/examples/forward/BUILD:85:8: Compiling examples/forward/example-forward-gradient-derivatives-using-eigen-with-scalar.cpp failed: 1 input file(s) do not exist
ERROR: /Users/rmeli/Documents/git/software/autodiff/examples/forward/BUILD:85:8 Compiling examples/forward/example-forward-gradient-derivatives-using-eigen-with-scalar.cpp failed: 1 input file(s) do not exist
INFO: Elapsed time: 0.247s, Critical Path: 0.02s
INFO: 2 processes: 2 internal.
FAILED: Build did NOT complete successfully
```

Looking at
https://github.com/autodiff/autodiff/blob/64346ae42b4a82cdefb364d289ebe5479cadc12c/examples/forward/BUILD#L85-L90
it seems that the paths are not up to date, since there is no `example-forward-gradient-derivatives-using-eigen-with-scalar.cpp` in `autodiff/examples/forward/`.

Manually updating the `cc_test` entries solves the issue. I haven't yet figured out how to use Catch2 v3 with `bazel` (all examples I could find use Catch2 v2), but I was wondering if it is worth the effort if `bazel` integration is not maintained? Any thoughts/suggestions here?

**Comment by allanleal** [comment]
@RMeli , thanks a lot for investigating the current status of Bazel support in the project.

> but I was wondering if it is worth the effort if bazel integration is not maintained? Any thoughts/suggestions here?

This will depend on community help. Unfortunately I cannot maintain bazel support in autodiff (too much on my plate already). 

I'm pinging @patrickhart and @markisus who in the past made contributions in this regard. Guys, would you kindly help us here? It would be good to have bazel builds in the CI so that we know that it has not been broken with changes in the lib.

**Comment by allanleal** [comment]
@RMeli , all looks good to me. Can we merge this?

**Comment by RMeli** [comment]
>  Unfortunately I cannot maintain bazel support in autodiff (too much on my plate already).

I totally get it! Unfortunately, I don't have the knowledge nor the time to take care of that either.

> @RMeli , all looks good to me. Can we merge this?

I converted to a draft to investigate how to update `bazel`. If further divergence with `bazel` is not a concern here (since it's already not working, should I open a separate issue for the record?), I'm happy to merge.

File: docs/website/installation.md [hunk 3]
```diff
@@ -15,7 +15,7 @@ conda install conda-forge::autodiff
 This will install {{autodiff}} in the conda environment that is active at the
 moment (e.g., the default conda environment is named `base`).
 
-### Developement Environment
+### Development Environment
```
Comment (RMeli) [inline docs/website/installation.md:18]: Fix typo from PR #249

**Comment by allanleal** [comment]
@RMeli , thanks a lot for your contribution. 

>  I open a separate issue for the record?

Yes, please. That would be good.

**Pull request merged**